sdks/python/apache_beam/transforms/enrichment_handlers/bigquery.py

```diff
@@ -0,0 +1,265 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import List
+from typing import Mapping
+from typing import Optional
+from typing import Union
+
+from google.api_core.exceptions import BadRequest
+from google.cloud import bigquery
+
+import apache_beam as beam
+from apache_beam.pvalue import Row
+from apache_beam.transforms.enrichment import EnrichmentSourceHandler
+
+QueryFn = Callable[[beam.Row], str]
+ConditionValueFn = Callable[[beam.Row], List[Any]]
+
+
+def _validate_batch_query_fn(query_fn, min_batch_size, max_batch_size):
+  if query_fn and min_batch_size and max_batch_size:
+    raise ValueError(
+        "Please provide exactly one of `query_fn` or "
+        "(`min_batch_size` and `max_batch_size`)")
+
+
+def _validate_bigquery_metadata(
+    table_name, row_restriction_template, fields, condition_value_fn, query_fn):
+  if query_fn and bool(table_name or row_restriction_template or fields or
+                       condition_value_fn):
+    raise ValueError(
+        "Please provide either `query_fn` or the parameters "
+        "`table_name`, `row_restriction_template`, and `fields` "
+        "together.")
+  elif not query_fn and not (table_name and row_restriction_template and
+                             (fields or condition_value_fn)):
+    raise ValueError(
+        "Please provide either `query_fn` or the parameters "
+        "`table_name`, `row_restriction_template`, and"
+        "`fields/condition_value_fn` together.")
+  if not query_fn and ((fields and condition_value_fn) or
+                       (not fields and not condition_value_fn)):
+    raise ValueError(
+        "Please provide exactly one of `fields` or "
+        "`condition_value_fn`")
```

Comment:
I think this might read cleaner as:

```
if query_fn:
   if bool(table_name or row_restriction_template or fields or
                       condition_value_fn):
      raise ...
else:
   ...

Reply:
done